helix-term/src/application.rs

```diff
@@ -98,14 +98,16 @@ impl Application {
         });
         let syn_loader = std::sync::Arc::new(syntax::Loader::new(syn_loader_conf));
 
+        let config = Arc::new(ArcSwap::from_pointee(config));
         let mut editor = Editor::new(
             size,
             theme_loader.clone(),
             syn_loader.clone(),
-            config.editor.clone(),
+            ArcSwap::from_pointee(config.load().editor.clone())
```

Comment:
Ah right, Editor was getting a subset of the config... Let's have Editor store the whole config Arc<> as well so it shares the ArcSwap with the Application level, then add a `fn config()` getter on editor that calls `config.map(|c| c.editor)` to do a projection of that subfield https://docs.rs/arc-swap/latest/arc_swap/struct.ArcSwapAny.html#method.map

Reply:
I'm not sure what the lifetime on `map()` is like, we might be able to do that here and avoid the extra helper method

Reply:
Allowing the editor to store the Application config will also mean `helix-view` will have to have `helix-term` as a dependency. That isn't possible though since it will be a cyclic dependency.

I think ideally, we can follow the pattern described in the [projecting into configuration field](https://docs.rs/arc-swap/latest/arc_swap/docs/patterns/index.html#projecting-into-configuration-field) section. I couldn't quite get the code for this to work though but I can keep fiddling. I may have been half asleep when attempting it.

In the Editor's config, I tried adding:

```rust
#[derive(Debug)]
pub struct Editor {
  pub config: Box<dyn DynAccess<Config>>,
}
```

But I get an error [rustc(2077)](https://doc.rust-lang.org/error-index.html#E0277):

```
`(dyn DynAccess<Config> + 'static)` doesn't implement `std::fmt::Debug`
```

I'm not sure how to get this to play nicely with the `#[derive(Debug)]` macro.

Reply:
You can manually implement debug like this: https://github.com/helix-editor/helix/blob/c0dbd6dc3f147b46d4352ac78903a7a70741cfd9/helix-view/src/document.rs#L125-L148

tbh I don't find all these `Debug` implementations useful anyway (they were previously added in https://github.com/helix-editor/helix/pull/143) since both editor and doc have so many fields that you don't want to directly print them out.

Let's remove the `Debug` bound from `Editor`?

Reply:
I've removed the debug trait